contrib/hashicorp/vault/vault_test.go

```diff
@@ -0,0 +1,321 @@
+package vault
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/mocktracer"
+
+	"github.com/hashicorp/vault/api"
+	"github.com/stretchr/testify/assert"
+)
+
+const secretMountPath = "/ns1/ns2/secret"
+
+func TestClient(t *testing.T) {
+	mt := mocktracer.Start()
+	defer mt.Stop()
+
+	t.Run("ok", func(t *testing.T) {
+		assert := assert.New(t)
+		client, err := api.NewClient(&api.Config{HttpClient: NewHTTPClient()})
+		assert.NoError(err)
+		assert.NotNil(client)
+		assert.Nil(err)
+	})
+
+	t.Run("error", func(t *testing.T) {
+		assert := assert.New(t)
+		var config = &api.Config{
+			HttpClient: NewHTTPClient(),
+			Address:    "http://bad host.com",
+		}
+		client, err := api.NewClient(config)
+		assert.Nil(client)
+		assert.Error(err)
+	})
+}
+
+func setupServer(t *testing.T) (*httptest.Server, func()) {
+	storage := make(map[string]string)
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch r.Method {
+		case http.MethodPut:
+			slurp, err := ioutil.ReadAll(r.Body)
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer r.Body.Close()
+			storage[r.URL.Path] = string(slurp)
+			fmt.Fprintln(w, "{}")
+		case http.MethodGet:
+			val, ok := storage[r.URL.Path]
+			if !ok {
+				http.Error(w, "No data for key.", http.StatusNotFound)
+				return
+			}
+			secret := api.Secret{Data: make(map[string]interface{})}
+			json.Unmarshal([]byte(val), &secret.Data)
+			secretJSON, err := json.Marshal(secret)
+			if err != nil {
+				t.Fatal(err)
+			}
+			fmt.Fprintf(w, "%s\n", secretJSON)
+		}
+	}))
+	return ts, func() {
+		ts.Close()
+	}
+}
+
+func setupClient(ts *httptest.Server) (*api.Client, error) {
+	config := &api.Config{
+		HttpClient: NewHTTPClient(),
+		Address:    ts.URL,
+	}
+	client, err := api.NewClient(config)
+	if err != nil {
+		return nil, err
+	}
+	return client, nil
+}
+
+func TestNewHTTPClient(t *testing.T) {
+	ts, cleanupTs := setupServer(t)
+	defer cleanupTs()
+
+	client, err := setupClient(ts)
+	if err != nil {
+		t.Fatal(err)
+	}
+	testMountReadWrite(client, t)
+}
+
+func TestWrapHTTPClient(t *testing.T) {
+	ts, cleanupTs := setupServer(t)
+	defer cleanupTs()
+
+	httpClient := http.Client{}
+	config := &api.Config{
+		HttpClient: WrapHTTPClient(&httpClient),
+		Address:    ts.URL,
+	}
+	client, err := api.NewClient(config)
+	if err != nil {
+		t.Fatal(err)
+	}
+	client.SetToken("myroot")
+
+	testMountReadWrite(client, t)
+}
+
+func mountKV(c *api.Client, t *testing.T) func() {
```

Comment:
Since I had a hard time learning what this is, would it be worthwhile to add some documentation here?
```go
// mountKV mounts a new path into Vault of type KV and returns a function to unmount it.
```